Large refactoring and improvement to METAR parser

dopplershift opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/Unidata/MetPy/blob/main/CONTRIBUTING.md
-->

#### Description Of Changes
* [x] Parser was not dealing well with a lot of weather code fields (like 'FU' for smoke), so restructure just to look for sequences of any of those codes.
* [x] Add parsed visibility to output data (#1306)
* [x] Put remarks in output data (#1306)
* [x] Improve tests to check more fields
* Significantly refactored METAR parser to reduce duplicated code and simplify conversion of parsed fields to final values
* Fixed wind values when report contains them in m/s
* Test suite now has a bunch of METARs explicitly checked for all fields (and driving modifications to parser)
* Test suite also has some tests that check counts of parsed (non-NaN) fields
* Many tweaks to PEG file to increase the reports that the parser can handle
    * Oddities
    * Variable visibility (not output)
    * Full runway visibility (not output)
    * More weather codes
* Fixed our handling of plotting/mapping weather symbols to allow for alternates (e.g. rain vs. snow + TS) and "unknown precipitation" (UP) from automated stations
* Tweaked the value for quite a few weather codes based on the WMO handbook

With all that, most of the gains in lines of code are in the actual generated parser. Outside of that, `metar.py` got shorter by ~200 lines--and `test_metar.py` got longer by a similar amount. This is a good thing IMO.

<!--
Feel free to remove check-list items aren't relevant to your change

Please use keywords (e.g., Fixes, Closes) to create link to the issues or pull
requests you resolved, so that they will automatically be closed when your pull
request is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### Checklist

- [x] Closes #1951
- [x] Closes #1306
- [x] Tests added
- [x] Fully documented

dopplershift commented:
Almost removed 140 lines from `metar.py`, and that's while *adding* the visibility group.

dopplershift commented:
Interesting. Getting a couple failures in the declarative image tests due more values/fields being plotted with this PR.

kgoebber commented:
So I just ran the tests on this PR for the declarative image tests and it appears that there are two small changes in each image, likely the result of a METAR being now parsed corrected. In the `test_declarative_station_plot_fontsize` there is a station model in NH that how adds temperature, dew point, updates the cloud cover, and adds a weather symbol. In `test_declarative_sfc_obs_change_units` there is a single station at the west edge of the image in MN that is now being correctly parsed and an extra '6' is added to the image. I think this is a case where we should update the images as the new parsing is generating more information that is now being correctly plotted.

dopplershift commented:
@kgoebber Yeah that's what I saw as well. My plan was to try to grab at least the NH ob and add that as a test case for the parser.

VORTEXJeff commented:
I did quite a bit of testing on this last week, and found that, in addition to FU, the following codes also do not plot:
BLSN, DS, DZ, DU, GR (or a mix with other types like TSRAGR), GS (or a mix), PL (or mix with any other type), UP (or a mix with any other type), FC, PO, SG, RASN, SHSN, SQ, SS, VA. Additionally, if thunder is occurring with any other precipitation type except for rain, it does not plot (TSSN, TSFZRAPL). Also, it appears that "-TSRA" plots the symbol for TSRASN.

dopplershift commented:
@VORTEXJeff are those testing results with this branch or with MetPy 1.0.1?

If you'd like to test with this PR, you can install from it with `pip install git+https://github.com/Unidata/MetPy.git@refs/pull/1953/merge`.

Either way, if you have any examples of those failing METARs, I'd love to add them to our test suite. It'd be nice to stomp out as many of these issues as we can while I'm digging in this code.

akrherz commented:
Perhaps here nor there, but attached is a listing of 587 different present weather code combinations and an example METAR in my database since 1 July 2021.  These parse codes are a result of [https://github.com/python-metar/python-metar](python-metar) processing.

[wxcodes.txt](https://github.com/Unidata/MetPy/files/6803394/wxcodes.txt)

VORTEXJeff commented:
The results were all with the existing 1.0.1. I downloaded the branch and the primary error - that the METAR parser would stop when encountering many weather codes (e.g., FU, etc.) appears to be fixed. Thank you! There is a smaller, but separate, bug involving the mapping of some of the sensible weather codes to the integers needs for the weather symbol font. For example, TSRA (and -TSRA) maps to 95, which is TSRASN, but it should map to 95a, and TSSN (and -TSSN) should map to 95b. There is an identical problem with the +TS groups and code 97. A few of the mixed precip thunder types don't map to any symbol, including TSRAGR (any intensity), TSRASN (any), +/-TSUP, +TSFZRA, and +TSFZRAPL. Finally, TSUP plots a sandstorm symbol. @akrherz METAR list was a big help with this testing!

dopplershift commented:
Tweaked our docs to display the weather symbols to match the WMO tables--which revealed a glitch in the current one that was accidentally using the high cloud symbols for auto codes 0-4 (instead of having them blank):
WMO manned station:
<img width="324" alt="image" src="https://user-images.githubusercontent.com/221526/125363849-a0c1fa80-e32e-11eb-9cea-69ab565b0f4b.png">

MetPy `current_weather`:
<img width="395" alt="image" src="https://user-images.githubusercontent.com/221526/125364485-c7ccfc00-e32f-11eb-8a18-c584d8150da6.png">

WMO auto station:
<img width="324" alt="image" src="https://user-images.githubusercontent.com/221526/125363923-c3541380-e32e-11eb-99ae-ad7b8b2633ba.png">

MetPy `current_weather_auto`:
<img width="394" alt="image" src="https://user-images.githubusercontent.com/221526/125364554-f1862300-e32f-11eb-9eee-12be3eee201e.png">

dopplershift commented:
@VORTEXJeff Some of the issues you've seen *might* be improved by using `current_weather_auto` instead of `current_weather`?

This gets us to a challenge we have here. So here's our current problem: We parse the METAR present weather text (e.g. +SHRA, TSRA) into the WMO codes (0-99), which then maps "nicely" onto the glyphs in our font. For reasons that make no damned sense to me, the WMO has different sets of digits mapping to different weather types depending on whether it's an automated station or manual observation (see for example 55 manual vs. 53 auto). There's also this issue:
<img width="408" alt="image" src="https://user-images.githubusercontent.com/221526/125365214-175ff780-e331-11eb-8799-adae9c1a899b.png">
So our mapping to a single code doesn't permit a way to designate the proper "alternate". Also, the current implementation of `wx_code_to_numeric`/`wx_code_map` has to consistently target one or the other of `current_weather` or `current_weather_auto`, and I'm not sure it's doing that.

This original design dates back to 2016 before the METAR parser and it seemed like it would work nicely. Clearly, it's causing us some problems now that we want to work with METAR obs. Is there much value in the WMO codes here?

I'm thinking the right answer here is to make a direct mapping of the METAR text codes to the font glyph we want--that seems like the only way to avoid the inevitably lossy translation involving the integer codes. Options:
1. Make sure we can pass an array/list of strings (e.g. `['TSRA', '+SN', 'FC']`) to `plot_symbol`, create a mapper `metar_weather` (open to other names) to use like we do `current_weather`, mapping directly the string to our unicode character for the font.
2. Change the METAR parser to output the right font glyph code (using the same kind of mapping as in (1)) in the various `current_wx1_symbol` columns, or creating new columns. We could make `symbol_mapper` optional for `plot_symbol`, and it would just treat the integers given directly as custom unicode characters for the font. Would have to be careful not to break existing users if we re-used the column (assuming they're just using it to pass to MetPy `StationPlot`)

Thoughts @dcamron @kgoebber @akrherz @VORTEXJeff ?

dopplershift commented:
So I dug into what GEMPAK does. 😡 🔥  It takes the METAR wx codes and converts them into a triplet of its own internal values (0-80, stored together as a single integer, see [`PT_WNMT`](https://github.com/Unidata/gempak/blob/master/gempak/source/prmcnvlib/pt/ptwnmt.f)), which gets converted to its own internal string language, which gets converted into the WMO code for a "manned station" (see [`PR_NSYM`](https://github.com/Unidata/gempak/blob/master/gempak/source/prmcnvlib/pr/prnsym.f), [`PT_WCOD`](https://github.com/Unidata/gempak/blob/master/gempak/source/prmcnvlib/pt/ptwcod.f), and [`PT_WSYM`](https://github.com/Unidata/gempak/blob/master/gempak/source/prmcnvlib/pt/ptwsym.f)). These codes are what's [drawn](https://github.com/Unidata/gempak/blob/master/doc/User_guide/pdf/apxC.pdf) (page C-5), (coded in [`IWTHR`](https://github.com/Unidata/gempak/blob/master/gempak/source/device/draw/iwthr.f)). There's also a separate function [`PR_WMAO`](https://github.com/Unidata/gempak/blob/master/gempak/source/prmcnvlib/pr/prwmao.f) that GEMPAK can use to map the codes for an automated station into those for a "manned station".

If you look at that table on C-5, you'll see that it looks like the WMO ones I posted above, but has a "200" section, one symbol 201 for volcano, 202 for sea spray, and 203..."Unknown precipitation from an automatic station". A big question mark. Which looks absolutely nothing like the ones in the automatic table above. 🙄

I think maybe the right thing to do here is similar. Map everything we can to the correct WMO manual station code. For the "UP" (unknown precip type--and hopefully only those) we map to something else...maybe e.g. -41 so that the code value is correct, but the sign means "auto"? 🤷 Then modify our `present_weather` to understand those "special" codes and handle mapping those to the *real* unknown precip symbols we have in our collection, instead of some janky question mark.

That approach (augmented codes) could also work for encoding e.g. rain vs. snow on 93, 94, 95, and 97. GEMPAK does it by adding 10 to the code for the ones with snow.

I hate meteorology.

VORTEXJeff commented:
I second @akrherz on the impressive amount of digging here @dopplershift! My thoughts are to probably go with something similar to your option #1 above making sure to not to break things for existing users (as an FYI, I am using a modified version of a marriage of both Unidata surface map scripts, and the patch I downloaded yesterday returned an error when I tried to plot 'current_weather', but this was a quick enough fix since I am now plotting 'current_wx1_symbol' instead).

I also think that using additional integers to differentiate between TSRA, TSSN, etc., is a good idea since I don't see any value in retaining the existing WMO ambiguous symbology (I expect no less from the students in my synoptic class). And, to second @akrherz, I also don't see much (any) value in retaining different symbols for automated vs manual observations.
[StationData.pdf](https://github.com/Unidata/MetPy/files/6809905/StationData.pdf)

My end goal is to create an auto-updating surface station plot using MetPy to replace the ancient system that finally went caput last fall here http://ww2010.atmos.uiuc.edu/(Gh)/wx/surface.rxml (look under surface observations).

Thanks again for your continuing efforts on this.

kgoebber commented:
I looked briefly at this yesterday after @VORTEXJeff comment about using the alternative symbols and quickly thought our current solution was untenable to working. I think the best path is as @dopplershift outline as option 1 with the the form that you outlined after reviewing the GEMPAK code, and with only supporting a limited number of symbols from the automated as you outlined and was supported by @akrherz and @VORTEXJeff.

sgdecker commented:
I think part of the issue is that the WMO symbols are really defined in the context of the synoptic code, which nobody (in the US anyway) uses. METARs have been shoe-horned into the pre-existing table of symbols. Some of the WMO symbols for manual observations (such as 52) disappear from the auto_station table because an automated station cannot report "intermittent" precip. Either it's precipitating at the time of the ob or it isn't. Similarly, there is no METAR code (whether the ob is automated or not) for intermittent precip, so those symbols are never used.

[This document](https://www.aviationweather.gov/docs/metar/wxSymbols_anno2.pdf) is instructive. It doesn't 100% match what GEMPAK does, but it is close.

dopplershift commented:
So I ended up doing a bunch more clean up and refactoring. Not only have I added vis and remarks, fixed 'FU', I fixed up a bunch of the variations on some fields so that the parser successfully parses a bunch more reports. I pretty much sat there hammering on the list @akrherz sent over until I felt good about what it wasn't getting through (only ~4). Also fixed an issue parsing reports with winds in m/s.

Regarding the handling of mapping the codes to symbols, I ended up tweaking our `current_weather` symbol mapper and `wx_code_map`:
* Symbols >=100 map into (100-N) in the automated symbols. '+UP' and friends are mapped to the appropriate >100 value.
* Values >=1000 signal that an alternate symbol should be used, based on the thousands digit. Thus, 2095 indicates that the 2nd alternate for the 95 symbol should be used (in this case the snow variation of TS 95).
* I also tweaked a bunch of symbols that were mapping incorrectly, at least based on my interpretation of WMO 306.

None of the above details on the mapping should matter to most users. The typical flow of `parse_metar_file` -> `StationPlot`/`ObsPlot` will just work better now at displaying the correct symbol. Or at least it should. Further bug reports always welcome--they shouldn't trigger a massive refactor next time.

@sgdecker Thanks for that. I'm guessing @kgoebber used it or something related to that in making our code table. Fitting with the rest of this project, it of course disagrees with the WMO. Aviation (that's the boxes 36 thru 38):

<img width="504" alt="image" src="https://user-images.githubusercontent.com/221526/128472577-7a919424-766e-4250-89a8-8178269eb6ff.png">

WMO:
<img width="434" alt="image" src="https://user-images.githubusercontent.com/221526/128472722-89ea0d25-6645-4ff1-8f6d-4732a11c8945.png">

🙄

dopplershift commented:
I think we can ignore CodeClimate and Codacy here. They're only complaining about complexity, which IMO is still improved.

kgoebber commented:
I can confirm that I was using a WMO code table as I look at it on my pin-up board at my desk. D'oh. Thanks for this great work @dopplershift, this will be a much improved METAR parser and so happy to get visibility out of the parser now as well.

akrherz commented:
Wow @dopplershift, amazing.  So to add gas to the fire, here's a listing of 306 unique visibility values my ingest has found this year, in case you want to jam these through and see what fun falls out.  There's some obvious GIGO here.

[vsby.txt](https://github.com/Unidata/MetPy/files/6946255/vsby.txt)

dopplershift commented:
@akrherz I haven't had a chance to look, but was the GIGO with this PR?

akrherz commented:
> I haven't had a chance to look, but was the GIGO with this PR?

Sorry for the confusion.  That example listing was the parsed output from `python-metar` package, not `metpy`.

dopplershift commented:
Thanks @akrherz. You weren't kidding about garbage. 120SM? 3/16SM? 1/4M? UGH. Something to worry about in the future if we're so inclined.

eliteuser26 commented:
Great work. I didn't expect there was a lot of work to be done on this issue. Problems with Metar are that they have a variable number of weather groups wether it is a regular or a special weather observation. This makes it hard to decode sometimes. Also the mixture of precipitation and their order doesn't make it obvious to decipher.
